commit 9b3a70e21be556b740f589c9b0e3d3f69e738a02
Author: Alexander V. Nikolaev <avn@avnik.info>
Date:   Sat Feb 6 08:09:06 2016 +0200

    Patch dlopen() to allow direct paths to all required libs
    
    This patch is NixOS specific

diff --git a/cairocffi/__init__.py b/cairocffi/__init__.py
index 307d58c..639303c 100644
--- a/cairocffi/__init__.py
+++ b/cairocffi/__init__.py
@@ -9,6 +9,7 @@
 
 """
 
+import os
 import sys
 from ctypes.util import find_library
 from pathlib import Path
@@ -21,13 +22,25 @@ VERSION = __version__ = (Path(__file__).parent / 'VERSION').read_text().strip()
 version = '1.17.2'
 version_info = (1, 17, 2)
 
+# Use hardcoded soname, because ctypes.util use gcc/objdump which shouldn't be required for runtime
+_LIBS = {
+    'cairo': '@cairo@/lib/libcairo@ext@',
+    'glib-2.0': '@glib@/lib/libglib-2.0@ext@',
+    'gobject-2.0': '@glib@/lib/libgobject-2.0@ext@',
+    'gdk_pixbuf-2.0': '@gdk_pixbuf@/lib/libgdk_pixbuf-2.0@ext@',
+}
 
 def dlopen(ffi, library_names, filenames):
     """Try various names for the same library, for different platforms."""
     exceptions = []
 
     for library_name in library_names:
-        library_filename = find_library(library_name)
+        library_fullname = _LIBS.get(library_name, None)
+        # bypass find_library, if full name specified via _LIBS exists, and not contains @library@
+        if library_fullname and not "@" in library_fullname and os.path.exists(library_fullname):
+            library_filename = library_fullname
+        else:
+            library_filename = find_library(library_name)
         if library_filename:
             filenames = (library_filename,) + filenames
         else:
